hipercdecode.1
'\" t
'\"
'\"	DO NOT EDIT! This file is generated from hipercdecode.1in
'\"
.TH hipercdecode 1 "Sun Apr 25 16:22:30 2021" "hipercdecode 0.0"
'\"
'\"==========================================================================
'\"	STRINGS and MACROS
'\"==========================================================================
'\"
'\"	Define strings for special characters that nroff doesn't have
'\"
'\"	N.B. using .if cua to test for special characters did not work.
'\"
.if !'\*[.T]'ps' .ds ua up
.if  '\*[.T]'ps' .ds ua \(ua
.if !'\*[.T]'ps' .ds da down
.if  '\*[.T]'ps' .ds da \(da
.if !'\*[.T]'ps' .ds <- left
.if  '\*[.T]'ps' .ds <- \(<-
.if !'\*[.T]'ps' .ds -> right
.if  '\*[.T]'ps' .ds -> \(->
'\"
'\"     bx - box a Courier string for making keycaps
'\"
'\"	N.B. this mess is to make the line drawing come out only
'\"	if we are really generating postscript
'\"
.de bx
.ie !'\*[.T]'ps' \{\
.	RB [ \\$1 ]\\$2
.\}
.el \{\
.	ie !r ps4html \{\
'\" \(br\|\s-1\f(CB\\$1\fP\s+1\|\(br\l'|0\(rn'\l'|0\(ul'
.		ft CW
.		nr par*bxw \w'\\$1'+.4m 
\Z'\v'.25m'\D'l 0 -1m'\D'l \\n[par*bxw]u 0'\D'l 0 1m'\D'l -\\n[par*bxw]u 0''\
\Z'\h'.2m'\s-1\\$1\s+1'\
\h'\\n[par*bxw]u'\\$2
.		ft P
.	\}
.	el \{\
.		RB [ \\$1 ]\\$2
.	\}
.\}
..
'\"
'\" strings to set current color (null with old groff)
'\"
.if mred .ds red \m[red]
.if mgreen .ds green \m[green]
.if mblue .ds blue \m[blue]
.if mblack .ds black \m[black]
.if mblack .ds mP \mP
'\"
'\" fix for grotty + xterm. We call for orange, grotty outputs yellow,
'\" but xterm displays yellow as orange.  The cycle is complete.
'\"
.if n .defcolor orange rgb #ffff00
'\"
'\" color <color> - set the current color (ignores request with old groff)
'\"
.de color
.if mred \m[\\$1]\c
..
'\"
'\" colorword <color> <word> - colorize a word (ignored by old groff)
'\"
.de colorword
.ie m\\$1 \m[\\$1]\\$2\mP\c
.el \\$2\c
..
'\"
'\" colbox <fg> <bg> <word> - colorize a word in a filled box
'\"
.de colbox
.ie mred \M[\\$2]\
\v'+.167v'\
\D'P 0 -0.9v  \w'\\$3'u 0  0 +0.9v   -\w'\\$3'u 0'\
\v'-.167v'\
\m[\\$1]\\$3\mP\MP
.el \\$3\c
..
'\"
'\"	Macros for doing pdfmarks
'\"
.de specialps
.if  '\*[.T]'ps' \\k_\X'ps: \\$*'\h'|\\n_u'\c
..
'\"
'\" pdfmark PDFMARKCODE
'\"
.ds pdfmarks
.if d pdfmarks \{\
.de pdfmark
.	specialps exec [\\$1 pdfmark
..
'\"
'\" pdfdest LINKNAME
'\"
.de pdfdest
.pdfmark "/Dest /\\$1 /View [/XYZ -5 PL null] /DEST"
..
'\"
'\" pdfbookmark COUNT LINKNAME STRING
'\"
.de pdfbookmark
.   pdfmark "/View [/XYZ 44 730 1.0] /Count \\$1 /Dest /\\$2 /Title (\\$3) /OUT"
..
'\"
'\"	Define the SH and SS macros to save pdfmark information
'\"	in "arrays" of numbers and strings.
'\"
.if !r rr_n \{\
.nr rr_n 0 1
.am SH
.	nr rr_levels!\\n+[rr_n] 2
.	ds rr_labels!\\n[rr_n] \\$*
.	pdfdest Link\\n[rr_n]
..
.am SS
.	nr rr_levels!\\n+[rr_n] 3
.	ds rr_labels!\\n[rr_n] \\$*
.	pdfdest Link\\n[rr_n]
..
.\}
'\"
'\"	Called at the end of the document to generate the pdfmark outline
'\"
.de pdf_outline
.nr rr_levels!\\n+[rr_n] 1
.nr rr_i 0 1
.while \\n+[rr_i]<\\n[rr_n] \{\
.   nr rr_ip1 \\n[rr_i]+1
.   nr rr_count 0
.   if \\n[rr_levels!\\n[rr_ip1]]>\\n[rr_levels!\\n[rr_i]] \{\
.       nr rr_j \\n[rr_i] 1
.       while \\n+[rr_j]<\\n[rr_n] \{\
.           if \\n[rr_levels!\\n[rr_j]]<=\\n[rr_levels!\\n[rr_i]] \{\
.               break
.           \}
.           if \\n[rr_levels!\\n[rr_j]]==(\\n[rr_levels!\\n[rr_i]]+1) \{\
.               nr rr_count \\n[rr_count]+1
.           \}
.       \}
.   \}
.   ds hhh \\*[rr_labels!\\n[rr_i]]
.   pdfbookmark -\\n[rr_count] Link\\n[rr_i] "\\*[hhh]"
.\}
..
'\"
'\" Some postscript to make pdfmarks harmless on old interpreters...
'\"
.specialps "def /pdfmark where {pop} {userdict /pdfmark /cleartomark load put} ifelse"
'\"
'\" Force display of Bookmarks in Acrobat when document is viewed.
'\"
.pdfmark "[/PageMode /UseOutlines /Page 1 /View [/XYZ null null null] /DOCVIEW"
'\"
'\" Output the document info in pdfmarks
'\"
.pdfmark "\
	/Title (\*[an-title](\*[an-section])) \
	/Subject (\*[an-title] Manual Page) \
	/Author (Rick Richardson) \
	/Keywords (printing) \
	/Creator (groff \n(.x.\n(.y.\n(.Y -man) \
	/CreationDate (\*[an-extra1]) \
	/ModDate (\*[an-extra1]) \
	/DOCINFO"
\}
'\"
'\" The manual page name is only 1st level mark
'\"
.nr rr_levels!\n+[rr_n] 1
.ds rr_labels!\n[rr_n] \*[an-title](\*[an-section])
.pdfdest Link\n[rr_n]
'\"
'\"==========================================================================
'\"	MANUAL PAGE SOURCE
'\"==========================================================================
.SH NAME
hipercdecode \- Decode a HIPERC stream into human readable form.
.SH SYNOPSIS
.B hipercdecode
.RI [ options "] <" hiperc-file
.SH DESCRIPTION
.B hipercdecode
decodes a HIPERC stream into human readable form.
Uncompressed and JBIG formats are handled.
.P
An HIPERC stream is the printer language used by the
Oki Data C310dn, C3100, C3200n, C3250n, C3300n, C3400n, C5100n, C5250n, C5500n,
C5600, and the C5800n
printers.

.SH COMMAND LINE OPTIONS
These are the options that can appear on the command line.
.TP
.BI \-d\0 basename
Basename of .pbm file for saving decompressed planes.
.TP
.BI \-h
Print hex file offsets.
.TP
.BI \-o
Print file offsets.
.TP
.BI \-D\0 level
Set Debug level [0].

.SH EXAMPLES
Decode an HIPERC stream file created by foo2hiperc.

.nf
.ft CW
$ foo2hiperc-wrapper testpage.ps | hipercdecode -h
     0:	\\033%-12345X@PJL
     f:	@PJL RDYMSG DISPLAY = "Unknown"
    30:	@PJL SET OKIJOBACCOUNTJOB USERID="Unknown" JOBNAME="Unknown"
    6e:	@PJL SET OKIAUXJOBINFO DATA="DocumentName=Unknown"
    a2:	@PJL SET OKIAUXJOBINFO DATA="ComputerName=dual.rkkda.org"
    dd:	@PJL SET OKIAUXJOBINFO DATA="ReceptionTime=00:00:00 2008/01/30"
   11e:	@PJL SET OKIAUTOTRAYSWITCH=ON
   13d:	@PJL SET OKIPAPERSIZECHECK=ENABLE
   160:	@PJL SET RESOLUTION=600
   179:	@PJL SET PAPER=LETTER
   190:	@PJL SET OKITRAYSEQUENCE=PAPERFEEDTRAY
   1b8:	@PJL SET OKIPAPERFEED=TRAY1
   1d5:	@PJL SET OKIMEDIATYPE = PLAIN
   1f4:	@PJL SET LPARM:PCL OKIPRINTMARGIN=INCH1D6
   21f:	@PJL SET COPIES=1
   232:	@PJL SET QTY=1
   242:	@PJL SET HIPERCEFFECTIVEBLOCKSIZE=34799360
   26e:	@PJL ENTER LANGUAGE=HIPERC
   289:	RECTYPE 0 (len=52,0x34 cnt=1)
   291:		BLKNUM 0, nbie=1, pn=3 [black] uc=0,0, wid=4864 ud=0,100
   2a5:		BLKNUM 1 (len=20), uncompressed=1, bie:
		DL = 48, D = 48, P = 49, - = 48, XY = 4864 x 6816
		L0 = 256, MX = 0, MY = 0
		Order   = 0 
		Options = 0 
		1 stripes, 0 layers, 49 planes
   2bd:	RECTYPE 1 (len=155668,0x26014 cnt=1)
   2c5:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
   2cd:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
 262d1:	RECTYPE 1 (len=155668,0x26014 cnt=2)
 262d9:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
 262e1:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
 4c2e5:	RECTYPE 1 (len=155668,0x26014 cnt=3)
 4c2ed:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
 4c2f5:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
 722f9:	RECTYPE 1 (len=155668,0x26014 cnt=4)
 72301:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
 72309:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
 9830d:	RECTYPE 1 (len=155668,0x26014 cnt=5)
 98315:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
 9831d:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
 be321:	RECTYPE 1 (len=155668,0x26014 cnt=6)
 be329:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
 be331:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
 e4335:	RECTYPE 1 (len=155668,0x26014 cnt=7)
 e433d:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
 e4345:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
10a349:	RECTYPE 1 (len=155668,0x26014 cnt=8)
10a351:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
10a359:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
13035d:	RECTYPE 1 (len=155668,0x26014 cnt=9)
130365:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
13036d:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
156371:	RECTYPE 1 (len=155668,0x26014 cnt=10)
156379:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
156381:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
17c385:	RECTYPE 1 (len=155668,0x26014 cnt=11)
17c38d:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
17c395:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
1a2399:	RECTYPE 1 (len=155668,0x26014 cnt=12)
1a23a1:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
1a23a9:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
1c83ad:	RECTYPE 1 (len=155668,0x26014 cnt=13)
1c83b5:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
1c83bd:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
1ee3c1:	RECTYPE 1 (len=155668,0x26014 cnt=14)
1ee3c9:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
1ee3d1:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
2143d5:	RECTYPE 1 (len=155668,0x26014 cnt=15)
2143dd:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
2143e5:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
23a3e9:	RECTYPE 1 (len=155668,0x26014 cnt=16)
23a3f1:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
23a3f9:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
2603fd:	RECTYPE 1 (len=155668,0x26014 cnt=17)
260405:		BLKNUM 0 (len=4), plane=3, uc=0,0,0
26040d:		BLKNUM 1 (len=155648), Data=00 00 00 00 00 00 00 00 00 00 00 ...
286411:	RECTYPE 1 (len=155668,0x26014 cnt=18)
286419:		BLKNUM 0 (len=4), plane=3, uc=0,0,0

ChangeLog
2020-06-10	Rick Richardson <rick.richardson@comcast.net>
	* osx-hotplug: Add PREFIX=/usr or PREFIX=/usr/local to cc rules

2020-05-05:	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: SaveToner implemented for color pages,
	  thanks to bedi@3mtech.sk .

2020-04-26:	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: Bug with SaveToner

2020-02-07:	Rick Richardson <rick.richardson@comcast.net>
	* Add Cassette 2 to foo2oak-z1-InputSlot.xml (KM-1635)

2019-09-09:	Rick Richardson <rick.richardson@comcast.net>
	* Add SmartInstallDisable-Tool.run since it was gone from
	      http://hplipopensource.com/hplip-web/smartinstall/
	  See HP SMART INSTALL NOTES in the INSTALL doc.

2019-04-13:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: Add install-aa to disable AppArmor on SUSE

2018-09-10:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: target ppd:	revert the 09-06 change.

2018-09-06:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: target ppd:
	    Use the older foomatic-ppdfile which only has postscript,
	    not postscript and pdf, for making ppd's.

2017-12-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printers: HP LaserJet Pro M12a, M12w

2017-03-20	Rick Richardson <rick.richardson@comcast.net>
	* Add ddstdecode, for decoding some Ricoh printers.

2016-10-21	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: install-hotplug-prog
	    Add /lib/systemd/systemd-udevd to the list for Ubuntu 16.

2016-09-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox Phaser 3040 using the foo2hbpl2 driver.

2016-08-30	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox Phaser 3010 using the foo2hbpl2 driver.

2016-03-13	Rick Richardson <rick.richardson@comcast.net>
	* icc2ps/lcms.h: USE_CUSTOM_SWAB defined.

2015-10-24	Rick Richardson <rick.richardson@comcast.net>
	* foomatic-db/printer/Epson-AcuLaser_CX17NF.xml:
		Change margins on Epson AcuLaser CX17NF.

2015-10-11	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Epson AcuLaser CX17NF using the foo2hbpl2 driver.

2015-07-04	pavel.shvagirev	<pavel.shvagirev@gmail.com>
	* Modify: hplj10XX to not call us twice on Ubuntu 14.

2015-05-11	Rick Richardson <rick.richardson@comcast.net>
	* Bug in foomatic-db/printer/Xerox-Phaser_6115MFP.xml . Should use
	  foo2lava driver. Thanks to user "ophers".

2015-02-11	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Oki C511 using foo2hiperc.

2014-12-21	Rick Richardson <rick.richardson@comcast.net>
	* osx-hotplug: make it work with OS X 10.10

2014-12-04	Rick Richardson <rick.richardson@comcast.net>
	* make cups: add "Sandboxing Relaxed" to cups-files.conf iff cups >= 2

2014-09-25	Rick Richardson <rick.richardson@comcast.net>
	* foo2hblp2, foo2hblp2-wrapper, foomatic et al:
	    Add paper size A5.
	* foo2hblp2 foomatic db entries:
	    Changed resolution to 600x600, 1200x600, and 1200x1200.
	    Note: need to recreate the printers if you want more choices!

2014-05-19	Rick Richardson <rick.richardson@comcast.net>
	* HP CLJ CP1215:  Adjust margins: *HWMargins: 18 14.40 18 14

2014-05-04	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2-wrapper: Change PAPER=21 to PAPER=0 for custom page sizes.

2014-04-17	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Fuji Xerox DocuPrint CM215 MFP using the foo2hbpl2
	driver.

2014-04-08	Rick Richardson <rick.richardson@comcast.net>
	* Upgrade to JBIGKIT 2.1

2014-04-01	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2 foomatic-db files and PPDs:
	    Bug with default PageSize.
	    PLEASE delete and recreate the foo2hbpl2 printers.

2014-03-22	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: another redundancy from Dave Coffin

2014-03-18	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: More envelope sizes.

2014-03-17	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: Bug, was using A4(1) instead of letter(4) by default
	* hbpldecode: Print out paper codes.

2014-03-15	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: More changes from Dave.  Still gets 
	    Segmentation fault (core dumped)
	* hbpldecode: Changed unsigned to int in 2 places.  Works good!!!

2014-03-13	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: Bug with no -d/-r.

2014-03-13	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: Changes to pick apart version 1 from
	    Dave Coffin <dcoffin@cybercom.net>
	* I get:
	    Decoding page 1 to page-01.ppm ...
	    Segmentation fault (core dumped)

2014-03-02	Rick Richardson <rick.richardson@comcast.net>

hipercdecode.c
/*
 * $Id: hipercdecode.c,v 1.8 2014/01/24 19:25:47 rick Exp $
 */

/*b
 * Copyright (C) 2003-2006  Rick Richardson
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * Author: Rick Richardson <rick.richardson@comcast.net>
b*/

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>

#include "hiperc.h"
#include "jbig.h"

/*
 * Global option flags
 */
int	Debug = 0;
char	*RawFile;
char	*DecFile;
int	PrintOffset = 0;
int	PrintHexOffset = 0;
FILE	*FpRaw[4];
FILE	*FpDec[4];

void
debug(int level, char *fmt, ...)
{
    va_list ap;

    if (Debug < level)
	return;

    va_start(ap, fmt);
    vfprintf(stderr, fmt, ap);
    va_end(ap);
}

int
error(int fatal, char *fmt, ...)
{
	va_list ap;

	fprintf(stderr, fatal ? "Error: " : "Warning: ");
	if (errno)
	    fprintf(stderr, "%s: ", strerror(errno));
	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	if (fatal > 0)
	    exit(fatal);
	else
	{
	    errno = 0;
	    return (fatal);
	}
}

void
usage(void)
{
    fprintf(stderr,
"Usage:\n"
"	hipercdecode [options] < oki-file\n"
"\n"
"	Decode a HIPERC into human readable form.  Uncompressed and JBIG\n"
"	are handled.\n"
"\n"
"	A HIPERC stream is the printer language used by the Oki Data C310dn,\n"
"	C3100, C3200n, C3250n, C3300n, C3400n, C5100n, C5250n, C5500n, C5600,\n"
"	and the C5800n printers.\n"
"\n"
"Options:\n"
"       -d basename Basename of .pbm file for saving decompressed planes\n"
"       -r basename Basename of .jbg file for saving raw planes\n"
"       -o          Print file offsets\n"
"       -h          Print hex file offsets\n"
"       -D lvl      Set Debug level [%d]\n"
    , Debug
    );

    exit(1);
}

static int
getdword(unsigned char buf[4])
{
    return (buf[0] << 24) | (buf[1] << 16) | (buf[2] << 8) | (buf[3] << 0);
}

void
print_bih(unsigned char bih[20])
{
    unsigned int xd, yd, l0;

    xd = (bih[4] << 24) | (bih[5] << 16) | (bih[6] << 8) | (bih[7] << 0);
    yd = (bih[8] << 24) | (bih[9] << 16) | (bih[10] << 8) | (bih[11] << 0);
    l0 = (bih[12] << 24) | (bih[13] << 16) | (bih[14] << 8) | (bih[15] << 0);

    printf("		DL = %d, D = %d, P = %d, - = %d, XY = %d x %d\n",
	 bih[0], bih[1], bih[2], bih[3], xd, yd);

    printf("		L0 = %d, MX = %d, MY = %d\n",
	 l0, bih[16], bih[17]);

    printf("		Order   = %d %s%s%s%s%s\n", bih[18],
	bih[18] & JBG_HITOLO ? " HITOLO" : "",
	bih[18] & JBG_SEQ ? " SEQ" : "",
	bih[18] & JBG_ILEAVE ? " ILEAVE" : "",
	bih[18] & JBG_SMID ? " SMID" : "",
	bih[18] & 0xf0 ? " other" : "");

    printf("		Options = %d %s%s%s%s%s%s%s%s\n", bih[19],
	bih[19] & JBG_LRLTWO ? " LRLTWO" : "",
	bih[19] & JBG_VLENGTH ? " VLENGTH" : "",
	bih[19] & JBG_TPDON ? " TPDON" : "",
	bih[19] & JBG_TPBON ? " TPBON" : "",
	bih[19] & JBG_DPON ? " DPON" : "",
	bih[19] & JBG_DPPRIV ? " DPPRIV" : "",
	bih[19] & JBG_DPLAST ? " DPLAST" : "",
	bih[19] & 0x80 ? " other" : "");
    printf("		%u stripes, %d layers, %d planes\n",
	((yd >> bih[1]) +  ((((1UL << bih[1]) - 1) & xd) != 0) + l0 - 1) / l0,
	bih[1] - bih[0], bih[2]);
}
	    unsigned char blk[50*1024*1024];

void
decode(FILE *fp)
{
    int		c;
    int		rc;
    FILE	*dfp = NULL;
    int		pageNum = 1;
    int		len;
    int		curOff = 0;
    struct jbg_dec_state	s[5];
    unsigned char	bih[4][20];
    int			imageCnt[4] = {0,0,0,0};
    int         	pn = 0;
    char		buf[1024];
    int			reccnt = 0;
    int			uncompressed = 0;
    unsigned int	w = 0, h = 0;

    c = getc(fp);
    if (c == EOF)
    {
	printf("EOF on file\n");
	return;
    }
    ungetc(c, fp);
    if (c == '\033')
    {
	while (fgets(buf, sizeof(buf), fp))
	{
	    if (PrintOffset)
		printf("%d:	", curOff);
	    else if (PrintHexOffset)
		printf("%6x:	", curOff);
	    if (buf[0] == '\033')
	    {
		printf("\\033");
		fputs(buf+1, stdout);
	    }
	    else
		fputs(buf, stdout);
	    curOff += strlen(buf);
	    if (strcmp(buf, "@PJL ENTER LANGUAGE=HIPERC\n") == 0)
		break;
	}
    }

    for (;;)
    {
	unsigned int	reclen;
	unsigned int	rectype;
	unsigned int	blknum;

	rc = fread(&reclen, 4, 1, fp);
	if (rc != 1) break;
	rc = fread(&rectype, 4, 1, fp);
	if (rc != 1) break;

	reclen = be32(reclen);
	rectype = be32(rectype);
	if (PrintOffset) printf("%d:	", curOff);
	else if (PrintHexOffset) printf("%6x:	", curOff);
	printf("RECTYPE %d (len=%d,0x%x cnt=%d)",
	    rectype, reclen, reclen, ++reccnt);
	if (rectype == 0)
	    printf(", Page %d\n", pageNum);
	else
	    printf("\n");
	curOff += 8;

	blknum = 0;
	reclen -= 8;
	while (reclen)
	{
	    int	blklen;
	    int	i;
//	    unsigned char blk[50*1024*1024];

	    if (PrintOffset) printf("%d:	", curOff);
	    else if (PrintHexOffset) printf("%6x:	", curOff);

	    rc = fread(&blklen, 4, 1, fp);
	    if (rc != 1) break;
	    curOff += 4;
	    reclen -= 4;
	    
	    blklen = be32(blklen);
	    rc = fread(&blk, blklen, 1, fp);
	    if (rc != 1) return;
	    curOff += blklen;
	    reclen -= blklen;


	    switch (rectype)
	    {
	    case 0:
		reccnt = 0;
		switch (blknum)
		{
		case 0:
		    printf("\tBLKNUM %d, nbie=%d, pn=%d",
			blknum, blk[0], blk[1]);
		    switch (pn = blk[1])
		    {
		    case 0:		printf(" [cyan] "); break;
		    case 1:		printf(" [magen]"); break;
		    case 2:		printf(" [yello]"); break;
		    case 3:		printf(" [black]"); break;
		    }
		    printf(" uc=%d,%d, wid=%d ud=%d,%x\n",
			blk[2], blk[3],
			getdword(blk+4), getdword(blk+8), getdword(blk+12));
		    w = getdword(blk+4);
		    if (pn < 0 || pn > 3)
			error(1, "Plane %d is not 0-3.\n", pn);
		    break;
		case 1:	/* BIH */
		    uncompressed = (blk[0] == 48);
		    printf("\tBLKNUM %d (len=%d), uncompressed=%d, bie:\n",
			blknum, blklen, uncompressed);
		    memcpy(bih[pn], blk, 20);
		    // bih[18] |= JBG_ILEAVE | JBG_SMID;
		    // bih[19] |= JBG_TPDON | JBG_DPON;
		    // bih[19] |= JBG_DELAY_AT;
		    print_bih(bih[pn]);
		    h = (blk[8] << 24) | (blk[9] << 16)
			| (blk[10] << 8) | (blk[11] << 0);
		    imageCnt[pn] = 0;
		    break;
		}
		break;
	    case 1:
		switch (blknum)
		{
		case 0:	/* PLANE */
		    printf("\tBLKNUM %d (len=%d), plane=%d, uc=%d,%d,%d\n",
			blknum, blklen, pn = blk[0], blk[1], blk[2], blk[3]);
		    break;
		case 1:	/* DATA */
		    printf("\tBLKNUM %d (len=%d), Data=", blknum, blklen);
		    for (i = 0; i < blklen && i < 11; ++i)
			printf("%02x ", blk[i]);
		    if (blklen >= 11)
			printf("...");
		    printf("\n");
		    if (RawFile && !FpRaw[pn])
		    {
			sprintf(buf, "%s-%02d-%d.jbg",
					RawFile, pageNum, pn);
			FpRaw[pn] = fopen(buf, "w");
			rc = fwrite(bih[pn], 1, 20, FpRaw[pn]);
		    }
		    if (FpRaw[pn])
			rc = fwrite(blk, 1, blklen, FpRaw[pn]);
		    if (uncompressed)
		    {
			if (DecFile && !FpDec[pn])
			{
			    sprintf(buf, "%s-%02d-%d.pbm",
					    DecFile, pageNum, pn);
			    FpDec[pn] = fopen(buf, "w");
			    fprintf(FpDec[pn], "P4\n");
			    fprintf(FpDec[pn], "%9d %9d\n", w, h);
			}
			if (FpDec[pn])
			    rc = fwrite(blk, 1, blklen, FpDec[pn]);
		    }
		    else if (DecFile)
		    {
			size_t		cnt;
			unsigned char	*image;

			if (imageCnt[pn] == 0)
			{
			    jbg_dec_init(&s[pn]);
			    rc = jbg_dec_in(&s[pn], bih[pn], 20, NULL);
			    if (rc == JBG_EIMPL)
				error(1, "JBIG uses unimplemented feature\n");
			    // rc = fwrite(bih, 20, 1, dfp);
			    imageCnt[pn] = 20;
			    len = jbg_dec_getsize(&s[pn]);
			    image = jbg_dec_getimage(&s[pn], 0);
			    if (image && len)
				memset(image, 0, len);
			}
			for (i = 0; i < blklen; ++i)
			{
			    rc = jbg_dec_in(&s[pn], &blk[i], 1, &cnt);
			    if (rc == JBG_EOK)
				break;
			    if (rc != JBG_EAGAIN && rc != JBG_EOK)
				error(1, "jbg_dec_in c=%x i=%d rc=%d (%s)\n",
				    blk[i], i, rc, jbg_strerror(rc));
			}
			if (0) {
			    printf("\ti=%d (%s)\n", i,  jbg_strerror(rc));
			    len = jbg_dec_getsize(&s[pn]);
			    image = jbg_dec_getimage(&s[pn], 0);
			    printf("\tlen=%d image=%lx (%s)\n",
				len, (long) image, jbg_strerror(rc));
			    for (i=0; i<len; ++i)
				if (image[i])
				    printf("image[%d] = %x\n", i, image[i]);
			}
			if (rc == JBG_EOK)
			{
			    int	w, h;

			    h = jbg_dec_getheight(&s[pn]);
			    w = jbg_dec_getwidth(&s[pn]);
			    len = jbg_dec_getsize(&s[pn]);
			    image = jbg_dec_getimage(&s[pn], 0);
			    printf("\trc=%d(%s) %dx%d len=%d\n",
				rc, jbg_strerror(rc), w, h, len);

			    sprintf(buf, "%s-%02d-%d.pbm",
					DecFile, pageNum, pn);
			    dfp = fopen(buf, "w");
			    if (!dfp)
				error(1, "Couldn't open '%s'.\n", buf);
			    fprintf(dfp, "P4\n%8d %8d\n", w, h);
			    rc = fwrite(image, 1, len, dfp);
			    fclose(dfp);

			    jbg_dec_free(&s[pn]);
			    imageCnt[pn] = 0;
			}
		    }
		    break;
		}
		break;
	    default:
		printf("\t");
		for (i = 0; i < blklen && i < 20; ++i)
		    printf("%02x ", blk[i]);
		if (blklen >= 22)
		    printf("...");
		printf("\n");
		break;
	    }

	    ++blknum;
	}
	if (rectype == 255)
	{
	    if (FpDec[pn])
	    {
		int	i;
		for (i = 0; i < 4; ++i)
		{
		    if (FpDec[i])
			fclose(FpDec[i]);
		    FpDec[i] = 0;
		}
	    }
	    c = getc(fp);
	    if (c == EOF) break;
	    ungetc(c, fp);
	    if (c == '\033') break;
	    ++pageNum;
	}
    }

    while (fgets(buf, sizeof(buf), fp))
    {
	if (PrintOffset)
	    printf("%d:	", curOff);
	else if (PrintHexOffset)
	    printf("%6x:	", curOff);
	if (buf[0] == '\033')
	{
	    printf("\\033");
	    fputs(buf+1, stdout);
	}
	else
	    fputs(buf, stdout);
	curOff += strlen(buf);
	if (strcmp(buf, "@PJL ENTER LANGUAGE=HIPERC\n") == 0)
	    break;
    }
    printf("\n");
}

int
main(int argc, char *argv[])
{
	extern int	optind;
	extern char	*optarg;
	int		c;

	while ( (c = getopt(argc, argv, "d:hor:D:?h")) != EOF)
		switch (c)
		{
		case 'd': DecFile = optarg; break;
		case 'r': RawFile = optarg; break;
		case 'o': PrintOffset = 1; break;
		case 'h': PrintHexOffset = 1; break;
		case 'D': Debug = atoi(optarg); break;
		default: usage(); exit(1);
		}

	argc -= optind;
	argv += optind;

	if (argc > 0)
	{
	    FILE	*fp;

	    fp = fopen(argv[0], "r");
	    if (!fp)
		error(1, "file '%s' doesn't exist\n", argv[0]);
	    decode(fp);
	    fclose(fp);
	}
	else
	    decode(stdin);

	exit(0);
}
